### ISSUE TITLE
Use a phf map for the WebGL extension registry

### ISSUE THREAD

**Issue by nox** [servo/servo#20357]
There is no need to build a HashMap at runtime etc, we could just generate it at compile-time.

**Comment by JacksonCoder** [comment]
I would like to take on this issue, if possible. What files or parts of the project handle the WebGL extension registry?

**Comment by JacksonCoder** [comment]
I think I found what parts of the application handle the HashMap of extensions: `components/script/dom/webgl_extensions/extensions.rs`, but I still can't find the code that initializes the HashMap, or at least not in a way that can be determined at compile time.

**Comment by jdm** [comment]
@JacksonCoder https://github.com/servo/servo/blob/199471231478425b0de3c29b473764dabfe45cba/components/script/dom/webgl_extensions/extensions.rs#L267-L276 is the code that fills the hashmap.

**Comment by JacksonCoder** [comment]
@jdm Thanks, I didn't see that. Do you think that the `rust-phf` crate (https://github.com/sfackler/rust-phf) should be the one we use to generate the compile-time map?

**Comment by jdm** [comment]
Yes, that's what we use in other places in the codebase too.

**Comment by JacksonCoder** [comment]
I'm almost finished, I'll make a PR either today or tomorrow. I tried to find an idiomatic way to create a list of types so I could could just generate the map in `build.rs` based on a single, constant list of types. Unforunately, the only crate I was able to find that could do that was `typemap`, and since I had to impl the Key trait for each type I wanted in the map, I decided it just was easier to just put an `entry` call manually for each type.

One problem I did encounter was the fact that `phf::Map` does not implement `JSTraceable` or `MallocSizeOf`, which means I'll have to custom-implement those traits for `WebGLExtensions`.

Should I make tests for the changes I have made?

**Comment by jdm** [comment]
The changes should be covered by the existing webgl tests.

**Comment by JacksonCoder** [comment]
I'm sorry for the wait.. I've been on vacation recently, and I haven't had the time to work on this. I forgot to mention one other problem I encountered: I need to import the extension types during the `build.rs` phase so I could generate the PHF map with the names of the extensions... however, I cannot use or import anything from the `script` crate because that would create a circular dependency. Right now, I'm seeing if I could use the alternative form of PHF map generation (macros).

**Comment by JacksonCoder** [comment]
The macro approach did work, but I still need to implement the traits that cannot be automatically derived due to the missing implementation on the PHF map type. My plan is to create trait implementations for the PHF map inside the crates of the trait implementations, which are within the servo repository, but outside of the `script` crate. Do you think that would be the best solution?

**Comment by jdm** [comment]
Are you referring to the [WebGLExtension](https://doc.servo.org/script/dom/webgl_extensions/extension/trait.WebGLExtension.html) trait, or some other type(s)? In general, I agree that moving types that you require at build time into a separate crate that `script` can depend upon would be the best solution. Moving `WebGLExtension` entirely won't work because of the `new` method which uses types which only exist in `script`, so that trait would need to be split up somehow if that were the case.

**Comment by jdm** [comment]
It would definitely be easiest to answer your questions if you pushed your code to a branch so it could be examined, and mentioned the error messages that appear.

**Comment by JacksonCoder** [comment]
You're right, I'll push the changes to my fork of the repo. I already fixed the problems with code generation (I just used inline macro generation instead), so I'm not worried about importing types at build time. My question (rephrased) was if I should implement the `JSTraceable` and `MallocSizeOf` traits manually on the PHF map type, which would allow me to auto-derive the `WebGLExtensions` struct.

**Comment by jdm** [comment]
Those would be useful implementations, yes!

**Comment by JacksonCoder** [comment]
Alright, I'll push to my fork once I finish implementing those.

**Comment by JacksonCoder** [comment]
I'm sorry, I haven't had much time to work on this recently. I did work on this a few days ago, and I found a few things: first of all, you have to use `phf-codegen` to get the names. What I did was create a map of names that was shared with the extensions and the map of extension names to their values. I also implemented the JSTraceable and MallocSizeOf traits.

**Comment by JacksonCoder** [comment]
Alright, so I did some more work today, and unfortunately I can say that moving this hash map over to a PHF map requires a lot of non-trivial, and some downright impossible things. First of all, I decided to make the map static, so it didn't need to be contained inside the `WebGLExtensions` struct (and therefore the trait impls were not needed). But then I ran into a problem: since the type we're holding as a value in the map is `Box<WebGLExtensionWrapper>`, it can't implement Copy, because a `Box` has a destructor. However, the PHF map stores it values internally as a `&'static [V]` map, where `V` is the value type. A static array slice of values can only be generated if the values implement Copy, which `Box<WebGLExtensionWrapper>` does not. So not only is a PHF map using non-Copyable values not allowed, but generating keys is a huge hassle that will no doubt be confusing for people working with WebGL extensions in the future. For that, the best 'solution' I could find was to generate _another_ PHF map that is both shared with `build.rs` and the `ext` module, which contains a mapping of types names to their extension name. That way, the names can be used in the extension PHF map.
Honestly, I think that the mild performance gain we'd get from using a PHF map is not worth it. Instead, I was thinking we could use a `lazy_static!` HashMap. That way, we don't have to explicitly register the extensions. We wouldn't have to use the PHF map for extension names, either.

**Comment by jdm** [comment]
Yeah, that sounds much more complex than I anticipated; sorry! A lazy_static sounds like a reasonable compromise.

**Comment by JacksonCoder** [comment]
Alright, I've pushed the code: https://github.com/jacksoncoder/servo. I also added the `cascade!` macro as a dependency to make HashMap generation look a little bit more clean.

**Comment by JacksonCoder** [comment]
So essentially, the problem is that having a static hash map results in the possibility of dangling pointers, where data the extensions are keeping track of are being garbage collected because the GC doesn't know the hash table exists? And in order to fix it, we'll have to make the hash table thread-local and call `trace_rust_roots` manually for the `WebGLExtensions::trace` method? If so, the only problem I see is that I don't know if a thread-local lazy_static is possible, but I'll look into it.

**Comment by nox** [comment]
All I wanted to replace is the correspondance from `str` to the function that enables the corresponding extension, I was on PTO and now that I'm reading about this again I'm not sure what `lazy_static` and custom hooks have to do with that. :/

**Comment by JacksonCoder** [comment]
At this point, I think that creating a thread-local lazy_static HashMap would be nearly the same thing as just having the HashMap be part of the `WebGLExtensions` struct, but the latter would be much easier to maintain.
